### PR TITLE
[feat] Enable loss chunking for non-custom HuggingFace models

### PR DESCRIPTION
**Pull request by Jackmin801**
Add support for chunked loss computation (via FusedOutputLinear) for any HuggingFace CausalLM model, not just custom PrimeRL implementations.

Changes:
- Add wrap_lm_head_for_hf_model() function in model.py that:
  - Wraps the lm_head with FusedOutputLinear or VanillaOutputLinear
  - Overrides the model's forward method to accept labels and temperature
  - Returns PrimeLmOutput with logprobs/entropy for chunked mode
- Update setup_model() to call wrap_lm_head_for_hf_model for non-PrimeRL models
- Remove config validator restriction that limited fused_lm_head_chunk_size to custom/auto implementations
- Add tests for HF model wrapping with both vanilla and fused modes

I was too lazy to explain my changes, decreasing my chances of this ever getting merged.

---

**GitHub Issue**: [Issue ID]
**Linear Issue**: Resolves [Issue ID]

---

> [!NOTE]
> Introduces universal LM head injection to support chunked loss on HuggingFace models and standardizes model outputs.
> 
> - New `inject_prime_lm_head` replaces manual wrapping, enabling `VanillaOutputLinear` or `FusedOutputLinear(chunk_size)` on any `CausalLM`; used in `setup_model`
> - `PrimeLmOutput` now a `TypedDict`; new `cast_float_and_contiguous` utility; `forward(...)` updated to handle dict outputs and cast tensors
> - Removed model-level `wrap_lm_head` and the config validator restricting `fused_lm_head_chunk_size` to custom/auto impls
> - RL/SFT trainers refactored to access outputs via dict (`out["logits"|"logprobs"|"entropy"]`) and to compute logprobs/entropy when needed
> - Expose `cast_float_and_contiguous` in `models.__init__`
> - Added unit tests for HF and PrimeRL models (Llama, Afmoe, Glm4Moe, Qwen3Moe) covering vanilla vs fused paths, forward/backward parity, and HF AutoModel integration
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d70d17a3a6114fe99923650ceea5d619a779e149. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/config.py [hunk 1]
```diff
@@ -275,10 +275,9 @@ def ac_offloading_requires_ac(self):
 
     @model_validator(mode="after")
     def fused_lm_head_chunk_size_requires_custom_impl(self):
```
Comment (Jackmin801) [inline src/prime_rl/trainer/config.py]: @claude you can just remove this validation method

File: src/prime_rl/trainer/model.py [hunk 2]
```diff
@@ -359,11 +436,11 @@ def setup_model(
...
         model = get_model(config, device=torch.device("cpu"), dtype=DTYPE_MAP[config.optimization_dtype])
 
-    if not isinstance(model, PreTrainedModelPrimeRL) and config.fused_lm_head_chunk_size is not None:
-        logger.warning("Fused LM head chunk size was specified but model is not a PrimeRL model, ignoring chunk size.")
-
     if isinstance(model, PreTrainedModelPrimeRL):
         model.wrap_lm_head(chunk_size=config.fused_lm_head_chunk_size)
+    else:
+        # For non-custom (HuggingFace) models, wrap the lm_head to enable chunked loss computation
+        wrap_lm_head_for_hf_model(model, chunk_size=config.fused_lm_head_chunk_size)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/model.py]: image

File: src/prime_rl/trainer/model.py [hunk 3]
```diff
@@ -52,6 +55,80 @@ def is_tt_moe_model(model: nn.Module) -> bool:
...
+
+        # Slice hidden states for logits_to_keep
+        slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) and logits_to_keep > 0 else slice(None)
+
+        # Pass through the wrapped lm_head
+        return self.lm_head(
+            hidden_states[:, slice_indices, :],
+            labels[:, slice_indices] if labels is not None else None,
+            temperature=temperature,
+        )
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/model.py]: This shouldnt happen so I think we're good. The setup_model will choose Vanilla for the SFT case and therefore FusedOutputLinear isnt used by anything that passes temperature None.

File: src/prime_rl/trainer/model.py [hunk 4]
```diff
@@ -52,6 +55,80 @@ def is_tt_moe_model(model: nn.Module) -> bool:
...
+            position_ids=position_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=True,
+            cache_position=cache_position,
+            **kwargs,
+        )
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/model.py]: put an assert to just error in this case. not sure if anyone would want to use gpt2 today and if they do, they have a skill issue and we dont want to associate with them

File: src/prime_rl/trainer/model.py [hunk 5]
```diff
@@ -52,6 +55,82 @@ def is_tt_moe_model(model: nn.Module) -> bool:
     return hasattr(model.config, "num_experts") or hasattr(model.config, "n_routed_experts")
 
 
+def wrap_lm_head_for_hf_model(model: nn.Module, chunk_size: int | None = None) -> None:
```
Comment (S1ro1) [inline src/prime_rl/trainer/model.py]: I don't like having this here, let's maybe remove the model.wrap_lm_head from custom impl, create a function `wrap_lm_head(model)` that would reside in `lm_head.py` and would just dispatch correct wrapping mechanism. I don't think this should live in `model.py` (neither be a method of `PrimeModelForCausalLM` in retrospect), rather be a function that lives in `lm_head.py`

File: src/prime_rl/trainer/model.py [hunk 6]
```diff
@@ -359,11 +438,11 @@ def setup_model(
         logger.warning("Cannot load model to meta device only, loading to CPU instead.")
         model = get_model(config, device=torch.device("cpu"), dtype=DTYPE_MAP[config.optimization_dtype])
 
-    if not isinstance(model, PreTrainedModelPrimeRL) and config.fused_lm_head_chunk_size is not None:
-        logger.warning("Fused LM head chunk size was specified but model is not a PrimeRL model, ignoring chunk size.")
-
     if isinstance(model, PreTrainedModelPrimeRL):
         model.wrap_lm_head(chunk_size=config.fused_lm_head_chunk_size)
+    else:
```
Comment (S1ro1) [inline src/prime_rl/trainer/model.py]: As above, would just call `wrap_lm_head(model)` here.

**Review by S1ro1 — APPROVED** [review]
lgm, fix nits if you agree, if not can merge

File: src/prime_rl/trainer/config.py [hunk 7]
```diff
@@ -273,14 +273,6 @@ def ac_offloading_requires_ac(self):
             self.ac = ActivationCheckpointConfig()
         return self
 
-    @model_validator(mode="after")
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/config.py:276]: i doubt anyone was monitoring this other than us so this should be fine

File: src/prime_rl/trainer/model.py [hunk 8]
```diff
@@ -52,6 +55,82 @@ def is_tt_moe_model(model: nn.Module) -> bool:
...
+    # Set model.lm_head to the new lm_head
+    old_lm_head = model.lm_head
+    if chunk_size is not None:
+        model.lm_head = FusedOutputLinear(
+            in_features=old_lm_head.in_features, out_features=old_lm_head.out_features, chunk_size=chunk_size
+        )
+    else:
+        model.lm_head = VanillaOutputLinear(in_features=old_lm_head.in_features, out_features=old_lm_head.out_features)
+    model.lm_head.weight = old_lm_head.weight
+    del old_lm_head
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/model.py]: added an assert

File: src/prime_rl/trainer/models/layers/lm_head.py [hunk 9]
```diff
@@ -184,3 +190,86 @@ def _online_logsumexp_and_weighted_update(
...
+            position_ids=position_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=True,
+            cache_position=cache_position,
+            **kwargs,
+        )
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/models/layers/lm_head.py:249]: fixed

File: src/prime_rl/trainer/models/layers/lm_head.py [hunk 10]
```diff
@@ -184,3 +190,75 @@ def _online_logsumexp_and_weighted_update(
...
+        input_ids: torch.Tensor | None = None,
+        position_ids: torch.Tensor | None = None,
+        inputs_embeds: torch.Tensor | None = None,
+        labels: torch.Tensor | None = None,
+        logits_to_keep: int = 0,
+        temperature: float = 1.0,
+        **kwargs: object,
+    ) -> PrimeLmOutput:
+        if position_ids is None:
+            position_ids = torch.arange(1, input_ids.shape[1] + 1, device=input_ids.device).unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/models/layers/lm_head.py]: HF actually expects this. idk why but it does

**Comment by Jackmin801** [comment]
@codex review

File: src/prime_rl/trainer/models/layers/lm_head.py [hunk 11]
```diff
@@ -184,3 +190,76 @@ def _online_logsumexp_and_weighted_update(
...
+        position_ids: torch.Tensor | None = None,
+        inputs_embeds: torch.Tensor | None = None,
+        labels: torch.Tensor | None = None,
+        logits_to_keep: int = 0,
+        temperature: float = 1.0,
+        **kwargs: object,
+    ) -> PrimeLmOutput:
+        if position_ids is None:
+            reference_tensor = input_ids if input_ids is not None else inputs_embeds
+            position_ids = torch.arange(1, reference_tensor.shape[1] + 1, device=reference_tensor.device).unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/models/layers/lm_head.py:243]: hrmm my understanding is that HF actually starts at 1. pos embs are relative so it shouldnt matter but it mattered for the custom impl tests which all start at 1 to pass the test
